hy/cmdline.py

```diff
@@ -13,17 +13,196 @@
 import importlib
 import py_compile
 import runpy
-
+import traceback
 import astor.code_gen
 
 import hy
 from hy.lex import LexException, PrematureEndOfInput, mangle
 from hy.compiler import HyTypeError, hy_compile
-from hy.importer import hy_eval, hy_parse
+from hy.importer import hy_eval, hy_parse, hy_ast_compile_flags
 from hy.completer import completion, Completer
 from hy.macros import macro, require
 from hy.models import HyExpression, HyString, HySymbol
-from hy._compat import builtins, PY3, FileNotFoundError
+from hy._compat import builtins, PY3, FileNotFoundError, str_type
+
+
+class HyPdb(object):
+    """A contextmanager that patches `bdb` and `pdb` modules to make them parse
+    Hy.
+
+    In order to affect the interactive console environment when using `exec`,
+    we create custom versions of `[pb]db` methods that use specific namespace
+    dictionaries.  Also, they force use of the closure's global and local
+    values to slightly reduce the number of functions that need patches.
+    """
+
+    @staticmethod
+    def pdb_compile(src, filename='<stdin>',
+                    mode='single', module_name='cmdline'):
+        hy_tree = hy_parse(src + '\n')
+        ast_root = ast.Interactive if mode == 'single' else ast.Module
+        hy_ast = hy_compile(hy_tree, module_name,
+                            root=ast_root)
+        code = compile(hy_ast, filename, mode, hy_ast_compile_flags)
+        return code
+
+    def __init__(self, ctx_globals=None, ctx_locals=None):
+        self.ctx_globals = ctx_globals
+        self.ctx_locals = ctx_locals
+
+    def pdbpp_getval_or_undefined(self):
+        _pdb = self.pdb
+        def _pdbpp_getval_or_undefined(self, arg):
+            """This is just for `pdb++`"""
+            try:
+                code = HyPdb.pdb_compile(arg)
+                return eval(code, self.curframe.f_globals,
+                            self.curframe.f_locals)
+            except NameError:
+                return _pdb.undefined
+
+    def hy_pdb_default(self):
+        def _hy_pdb_default(self, line):
+            if line[:1] == '!': line = line[1:]
```

Comment:
I think the condition here is equivalent to `line.startswith('!')`, which seems like a clearer way to write it.

Reply:
All these patched functions are direct copies from the `[bp]db` source (a couple have slight cross-version modifications, though), but, yeah, they could be cleaned up a bit.

Reply:
In that case, you should attribute the source of the copied code in the commit message.